src/transformStateWithClones.js

```diff
@@ -7,7 +7,60 @@
  * @return {Object[]}
  */
 function transformStateWithClones(state, actions) {
-  // write code here
+  const stateArr = [];
+
+  for (const action of actions) {
+    const { type, extraData, keysToRemove } = action;
+    const newState = stateArr.length
+      ? { ...stateArr[stateArr.length - 1] }
+      : { ...state };
+
+    switch (type) {
+      case 'addProperties':
+        stateArr.push(Object.assign(newState, extraData));
+        break;
+
+      case 'removeProperties':
+        if (!keysToRemove.length) {
+          stateArr.push(newState);
+          break;
+        }
+
+        for (const key of keysToRemove) {
+          if (Object.hasOwn(newState, key)) {
+            delete newState[key];
+          }
+        }
+        stateArr.push(newState);
+        break;
+
+      case 'clear':
+        stateArr.push({});
+        break;
+    }
```

Comment:
Consider adding a `default` case in your `switch` statement. This is a good practice for error handling, as it can catch unexpected `type` values and handle them accordingly. [CODE STYLE]